lib/action_policy/behaviours/policy_for.rb

```diff
@@ -57,12 +57,20 @@ def implicit_authorization_target!
         )
       end
 
-      def policy_for_cache_key(record:, with: nil, namespace: nil, context: authorization_context, **)
+      def policy_for_cache_key(record:, with: nil, namespace: nil, context: nil, **)
         record_key = record._policy_cache_key(use_object_id: true)
-        context_key = context.values.map { _1._policy_cache_key(use_object_id: true) }.join(".")
+        context_key = context_for_policy(context).values.map { _1._policy_cache_key(use_object_id: true) }.join(".")
 
         "#{namespace}/#{with}/#{context_key}/#{record_key}"
       end
+
+      private def context_for_policy(context)
+        context ? authorization_context.merge(context) : memoized_authorization_context
+      end
+
+      private def memoized_authorization_context
+        @_authorization_context ||= authorization_context
+      end
```

Comment:
I moved memoization from `authorization_context` to this private method otherwise it could lead to nonintuitive issues.

My first solution was to merge the explicit context to a newly built one :
```ruby
context = context ? build_authorization_context.merge(context) : authorization_context
```

But then, an overrided `authorization_context` would be ignored:

```ruby
def authorization_context
  { application: current_application_name, user: current_user }
end

def allowed_to_do_something?
  user = find_user

  allowed_to? :do_something?, context: { user: user } 
  # Why application from my authorization_context is not merged ?
end
```